### PR TITLE
expose IAsyncEnumerable API for watcher

### PR DESCRIPTION
**Pull request by tg123**
#584
#581 

require C# 8.0 to use this API
`netstandard2.0` is supported with `Microsoft.Bcl.AsyncInterfaces`

_Note_: here `onError` to `CreateWatchEventEnumerator` is to handle non-fatal errors:
 * server yields a bad line, cannot be deserialized
 * server yields a `error` event

the enumerator will remain alive when `errros` above happens

### CODE REVIEW

File: src/KubernetesClient/WatcherExt.cs
Comment (tg123) [inline src/KubernetesClient/WatcherExt.cs:69]: doc will be added after API style get reviewed

**Comment by tg123** [comment]
cc @davidfowl @lodejard

File: src/KubernetesClient/Watcher.cs
Comment (davidfowl) [inline src/KubernetesClient/Watcher.cs:169]: This looks a bit convoluted. It's because StreamReader doesn't support cancellation token right?

File: src/KubernetesClient/Watcher.cs
Comment (davidfowl) [inline src/KubernetesClient/Watcher.cs:158]: What's the goal of this method?

File: src/KubernetesClient/Watcher.cs
Comment (davidfowl) [inline src/KubernetesClient/Watcher.cs:176]: Typically better to use `cancellationToken.ThrowIfCancellationRequested()`

File: src/KubernetesClient/Watcher.cs
Comment (tg123) [reply src/KubernetesClient/Watcher.cs:169]: disposing watcher will dispose the stream which would cancel the `readlineasync` in the original watcher pattern
both `streamReaderCreator()` and `readlineasync` does not honor the cancellation token, here is a workaround to attach it

File: src/KubernetesClient/Watcher.cs
Comment (tg123) [reply src/KubernetesClient/Watcher.cs:176]: agree 
but here is to be compatible with old watcher pattern
otherwise, signal a cancel will raise an exception.

**Comment by brendandburns** [comment]
I'm going to defer to @davidfowl and @lodejard to do the first reviews on this. Once they're happy I'll do a final review.

File: src/KubernetesClient/Watcher.cs
Comment (lodejard) [reply src/KubernetesClient/Watcher.cs:169]: Interesting, haven't seen this ContinueWith/CancellationToken pattern before. I'll look into that a little closer before commenting.

Comment (lodejard) [reply src/KubernetesClient/Watcher.cs:169]: Looking into this it seems the underlying task remains active even though the wrapping task becomes cancelled. I think you'd be okay without this pattern.

In a perfect world ReadLineAsync would take a CT... Short of that, to cover the `ReadLineAsync` it might be better to register the CT as a call to stream.Dispose. That will have the same effect as before - ensuring the underlying call waiting on network unwinds immediately.

```csharp
using var streamReader = await AttachCancellationToken(streamReaderCreator()).ConfigureAwait(false);
using var cancellationTokenRegistration = cancellationToken.Register(streamReader.Dispose);
```

For the first call down to content `ReadAsStreamAsync` --- that one doesn't take a CT either. Good news is it doesn't seem like that code path enters a wait state? In the original code it seems like that wasn't covered by cancellation either. So should be okay to leave the `streamReaderCreator()` unguarded by CT.

File: src/KubernetesClient/Watcher.cs
Comment (lodejard) [reply src/KubernetesClient/Watcher.cs:176]: For callers using `await foreach(...WatchAsync(CT))` it could be expected when a cancel is requested the await will throw? Might not be a mandatory convention, but it would indicate if the loop ended normally from response ending vs ending from a client-side cancellation teardown.

Ah - also the await keyword in the loop will also throw the exception if the ReadLineAsync task enters a cancelled state, so you'll need to catch for that as well for old watcher pattern.

How about if the `WatcherLoop` method has an additional `catch (OperationCancelledException)` case which does not call OnError? 

That would stop cancel exceptions from reaching the old pattern, but enable you to use `ThrowIfCancellationRequested` from the async enumerable function and allow `await foreach` callers to have task.Canceled status propagate back.

File: src/KubernetesClient/Watcher.cs
Comment (tg123) [reply src/KubernetesClient/Watcher.cs:169]: yup 
`streamReaderCreator` is under our control and easy to make it take CT

I am thinking that `AttachCancellationToken` is a waste if I use it only once :)

**Comment by tg123** [comment]
any idea about `Action<Exception> onError` parameter, I do not like the design

**Comment by mpoettgen** [comment]
@tg123,

> any idea about `Action<Exception> onError` parameter, I do not like the design

How about making `WatchEvent` a standalone, generic class `WatchEvent<T>` instead of a nested type and adding a third property `Error` of type `Exception`. `WatchAsync` should then return an `IEnumerable<WatchEvent<T>>` and you can either `yield return` the de-serialized watch event or a `WatchEvent` with `WatchEventType.Error` and the corresponding `Exception` in the `Error` property.

The other option would be to make the returned tuple a triple, but to me that would be a bit awkward.

**Comment by tg123** [comment]
ping @davidfowl @lodejard

**Comment by davidfowl** [comment]
Will take another look

**Comment by tg123** [comment]
netstandard2.0 support dropped

**Comment by davidfowl** [comment]
> netstandard2.0 support dropped

That's a big breaking change, not sure you should do that.

**Comment by tg123** [comment]
> > netstandard2.0 support dropped
> 
> That's a big breaking change, not sure you should do that.

netstarnd2 dropped here #603

File: src/KubernetesClient/Watcher.cs
Comment (mpoettgen) [inline src/KubernetesClient/Watcher.cs:194]: @tg123, not sure, but if I'm not mistaken, if the list/watch operation actually returns existing objects, then those will be read (again) by the first ReadLineAsync above. The `line` will not deserialize properly, as the line represents the object list, not a watch event. Therefore a `NullReferenceException` will be thrown, caught below and reported via the `onError` handler if and only if an `onError` handler has been provided. Test cases do not cover this. I don't think this is intentional!?

The `LineSeparatedHttpContent` only peeks at the first line, so it is read again here. It doesn't know, whether it should discard the line as it doesn't know, whether it is reading an object list or a watch event.

Comment (tg123) [reply src/KubernetesClient/Watcher.cs:194]: i noticed the npe during dev and did some changes to testcase https://github.com/kubernetes-client/csharp/pull/586/files/c8506d22c52caf3e827a51a9f4de477659557759#diff-b8e0f043f040cdeca7bd66470c6f4cf8dc42b39a1f514cf412e284309284e431L196

Yet, server will never return a `listobj` if `watch=true` and `LineSeparatedHttpContent` will install to `watch=true` only.

In the old days, (1.6 or 1.7 i believe), there is a `/watch/xxx` api which return `listobj` which will cause the issue you mentioned. but it was deprecated and `get` with `watch=true` will yield `watch obj` only
that is the reason i changed the testcase removing the first list obj.

Comment (mpoettgen) [reply src/KubernetesClient/Watcher.cs:194]: Then could you please have a look at the TestWatchWithHandlers? That's the test I stumbled on.

Comment (tg123) [reply src/KubernetesClient/Watcher.cs:194]: yup
`await WriteStreamLine(httpContext, MockKubeApiServer.MockPodResponse).ConfigureAwait(false);`
should be removed 
it is outdated, but i did not notice it as test passed

Comment (mpoettgen) [reply src/KubernetesClient/Watcher.cs:194]: That is because there is no `onError` handler attached. I'm experimenting with an alternative implementation that encapsulates both success and error results in a `WatchEvent<T>`. That's why I noticed.

**Comment by brendandburns** [comment]
Where is this PR at? I was waiting for the discussion to settle. Is it ready for a final review?

**Comment by brendandburns** [comment]
Checking in on the status of this PR again :)

**Comment by Woazboat** [comment]
Also interested in the status of this PR (since the current watcher api is extremely cumbersome to use)

**Review by davidfowl — APPROVED** [review]

**Comment by tg123** [comment]
talked with @davidfowl and got approval from him
cc @brendanburns

**Comment by tg123** [comment]
ping @brendanburns

**Comment by davidfowl** [comment]
image

The gray approval 😄

**Pull request merged**
